lib/streamlit/connections/snowflake_connection.py

```diff
@@ -162,7 +162,7 @@ def query(
         >>> df = conn.query("select * from pet_owners")
         >>> st.dataframe(df)
         """
-        from snowflake.connector import Error as SnowflakeError
+        from snowflake.connector.errors import TokenExpiredError  # type: ignore[import]
```

Comment:
@sfc-gh-mkeller quick question about the error types thrown from the `snowflake-python-connector` library:

Is there a specific error type that gets thrown in the situation where a connection's auth token has expired? Ideally, we'll want the Streamlit wrapper to automatically clear/reset the connection when this happens. Looking at `snowflake.connector.errors`, this error name/description seems like it's the right one, but I wasn't able to find it being thrown anywhere in the library.

Reply:
`TokenExpiredError` [doesn't seem to be ever used](https://github.com/search?q=repo%3Asnowflakedb%2Fsnowflake-connector-python%20TokenExpiredError&type=code)

Here's the order of how thing will be called if a master token expires:
https://github.com/snowflakedb/snowflake-connector-python/blob/a3b9f1142fe44c69f0ed330612e18d67b73202c8/src/snowflake/connector/network.py#L748C38-L748C38
https://github.com/snowflakedb/snowflake-connector-python
https://github.com/snowflakedb/snowflake-connector-python/blob/a3b9f1142fe44c69f0ed330612e18d67b73202c8/src/snowflake/connector/network.py#L523/blob/a3b9f1142fe44c69f0ed330612e18d67b73202c8/src/snowflake/connector/network.py#L527

https://github.com/snowflakedb/snowflake-connector-python/blob/a3b9f1142fe44c69f0ed330612e18d67b73202c8/src/snowflake/connector/network.py#L580
or
https://github.com/snowflakedb/snowflake-connector-python/blob/a3b9f1142fe44c69f0ed330612e18d67b73202c8/src/snowflake/connector/network.py#L587
So it should either be a [ProgrammingError](https://github.com/snowflakedb/snowflake-connector-python/blob/a3b9f1142fe44c69f0ed330612e18d67b73202c8/src/snowflake/connector/errors.py#L400) with `sqlstate=08001` or a [ReauthenticationRequest](https://github.com/snowflakedb/snowflake-connector-python/blob/a3b9f1142fe44c69f0ed330612e18d67b73202c8/src/snowflake/connector/network.py#L296) with `cause` being that aforementioned `ProgrammingError`.

`ReauthenticationRequest` will then be caught [here](https://github.com/snowflakedb/snowflake-connector-python/blob/a3b9f1142fe44c69f0ed330612e18d67b73202c8/src/snowflake/connector/network.py#L696) which is reraised for every connection type that isn't externalbrowser [here](https://github.com/snowflakedb/snowflake-connector-python/blob/a3b9f1142fe44c69f0ed330612e18d67b73202c8/src/snowflake/connector/network.py#L698).

Reply:
Thanks! Looks like we should be able to inspect the attrs of the error to narrow down the `ProgrammingErrors` that get raised to the ones that we want to retry for. Will make the changes to do that